### DISCUSSION TITLE
How to perform Pinch and Zoom on an Image element in Android device using Appium and W3C selenium apis

### DISCUSSION THREAD

**Discussion by AmanDaharwal** [appium/appium#18842] [Q&A]
We have image element in android device which get zoom on pinching, how to pinch and zoom the image element( android.widget.ImageView) using Appium and W3C selenium apis 

Used below code : 

`
// Define the center coordinates of the image
        int centerX = element.getLocation().getX() + element.getSize().getWidth() / 2;
        int centerY = element.getLocation().getY() + element.getSize().getHeight() / 2;

        // Define the scale factor for zooming
        double scaleFactor = 1.5; // You can adjust this value to control the amount of zoom

        PointerInput finger1 = new PointerInput(PointerInput.Kind.TOUCH, "finger1");
        PointerInput finger2 = new PointerInput(PointerInput.Kind.TOUCH, "finger2");

        Sequence pinchStart = new Sequence(finger1, 0);
        pinchStart.addAction(finger1.createPointerMove(Duration.ofMillis(0), PointerInput.Origin.viewport(), centerX, centerY))
                .addAction(finger2.createPointerMove(Duration.ofMillis(0), PointerInput.Origin.viewport(), centerX, centerY + 50))
                .addAction(finger1.createPointerDown(PointerInput.MouseButton.LEFT.asArg()))
                .addAction(finger2.createPointerDown(PointerInput.MouseButton.LEFT.asArg()));

        Sequence pinchMove = new Sequence(finger2, 1);
        pinchMove.addAction(finger1.createPointerMove(Duration.ofMillis(1000), PointerInput.Origin.viewport(), centerX, centerY))
                .addAction(finger2.createPointerMove(Duration.ofMillis(1000), PointerInput.Origin.viewport(), centerX, centerY + (int)(50 * scaleFactor)));

        Sequence pinchEnd = new Sequence(finger1, 2);
        pinchEnd.addAction(finger1.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));
        pinchEnd.addAction(finger2.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));

        // Perform the pinch and zoom action
        appiumDriver.perform(Arrays.asList(pinchStart, pinchMove, pinchEnd));
`

Getting below error - 
`org.openqa.selenium.InvalidArgumentException: The action id 'finger1' has one one or more duplicates
Build info: version: '4.10.0', revision: 'c14d967899'
System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '12.3', java.version: '17.0.5'`

**Comment by jlipps** [comment]
You're constructing the pinch incorrectly, as is indicated by the error message where you are using the same finger in two simultaneous sequences. You just want two Sequences. Each Sequence should have the following actions: move, pointer down, move, pointer up.

If you need to improve your conceptual understanding of how action sequences work, have a look at the selenium java client docs.
